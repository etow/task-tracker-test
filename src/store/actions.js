import taskRepository from '../repositories/taskRepository';
import taskCategoryRepository from '../repositories/taskCategoryRepository';
import {
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASKS,
  ADD_TASK,
  DELETE_TASK,
  SET_TASK_TO_EDIT,
  SET_TASK,
  FETCH_TASKS_CATEGORIES,
  FETCH_TASKS,
  CREATE_TASK,
  UPDATE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
} from './constants';

export default {
  [FETCH_TASKS_CATEGORIES]: ({ commit }) => {
    return taskCategoryRepository.get()
      .then((response) => {
        commit(SET_TASKS_CATEGORIES, response.data);
        commit(SET_TASKS_CATEGORIES_DATA, response.data);
      });
  },
  [FETCH_TASKS]: ({ commit }) => {
    return taskRepository.get()
      .then((response) => commit(SET_TASKS, response.data));
  },
  [CREATE_TASK]: ({ commit, state }, newTask) => {
    commit(BACKUP_STATE, state);
    const task = {
      ...newTask,
      id: Math.random().toString(36).substr(2, 6),
      order: state.tasks[newTask.category].length,
    };
    commit(ADD_TASK, task);
    return taskRepository.create(task);
  },
  [DELETE_TASK]: ({ commit, state }, {task, category}) => {
    commit(BACKUP_STATE, state);
    commit(DELETE_TASK, {task, category});
    commit(SET_TASK_TO_EDIT, {});
    return taskRepository.delete(task.id);
  },
  [UPDATE_TASK]: ({ commit, state }, {task, prevCategory}) => {
    commit(BACKUP_STATE, state);
    if (prevCategory && task.category !== prevCategory) {
      commit(DELETE_TASK, { task, category: prevCategory });
    }
    const targetIndex = state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    if (targetIndex >= 0) {
      commit(SET_TASK, { task, targetIndex });
    } else {
      const order = state.tasks[task.category].length;
      commit(ADD_TASK, { ...task, order });
    }
    commit(SET_TASK_TO_EDIT, {})
    return taskRepository.update(task);
  },
  [UPDATE_TASKS]: ({ commit, state }, { tasks, category }) => {
    commit(BACKUP_STATE, state);
    tasks.forEach((task, index) => {
      task.category = category;
      task.order = index;
    });
    commit(UPDATE_TASKS, { tasks, category });

    return taskRepository.update(tasks);
  },
  [SET_TASK_TO_EDIT]: ({ commit, state}, task) => {
    commit(BACKUP_STATE, state);
    commit(SET_TASK_TO_EDIT, task);
  },
  [ROLLBACK_STATE]: ({ commit }) => {
    commit(ROLLBACK_STATE);
  },
};
