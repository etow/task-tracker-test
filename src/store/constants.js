

/* mutations and actions */
export const ADD_TASK = 'ADD_TASK';
export const SET_TASKS = 'SET_TASKS';
export const SET_TASKS_CATEGORIES = 'SET_TASKS_CATEGORIES';
export const SET_TASKS_CATEGORIES_DATA = 'SET_TASKS_CATEGORIES_DATA';
export const SET_TASK = 'SET_TASK';
export const SET_TASK_TO_EDIT = 'SET_TASK_TO_EDIT';
export const DELETE_TASK = 'DELETE_TASK';
export const FETCH_TASKS_CATEGORIES = 'FETCH_TASKS_CATEGORIES';
export const FETCH_TASKS = 'FETCH_TASKS';
export const CREATE_TASK = 'CREATE_TASK';
export const UPDATE_TASK = 'UPDATE_TASK';
export const ROLLBACK_STATE = 'ROLLBACK_STATE';
export const BACKUP_STATE = 'BACKUP_STATE';
export const UPDATE_TASKS = 'UPDATE_TASKS';
export const UPDATE_TASK_ACTIVITY = 'UPDATE_TASK_ACTIVITY';
