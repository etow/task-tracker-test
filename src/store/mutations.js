import {
  ADD_TASK,
  SET_TASKS,
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASK,
  SET_TASK_TO_EDIT,
  DELETE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
} from './constants';

export default {
  [ADD_TASK]: (state, task) => {
    state.tasks[task.category].push(task);
  },
  [SET_TASKS]: (state, tasks) => {
    Object.keys(state.tasks)
      .forEach((category) => {
        const tasksByCategory = tasks.filter((task) => task.category === category);
        state.tasks[category] = tasksByCategory.sort((a, b) => a.order - b.order);
      });
  },
  [SET_TASKS_CATEGORIES]: (state, categories) => {
    categories.forEach((cat) => state.tasks[cat.name] = []);
  },
  [SET_TASKS_CATEGORIES_DATA]: (state, categories) => {
    categories.forEach((cat) => state.categories[cat.name] = cat);
  },
  [SET_TASK]: (state, { task, targetIndex }) => {
    state.tasks[task.category][targetIndex] = task;
  },
  [SET_TASK_TO_EDIT]: (state, task) => {
    state.taskToEdit = task;
  },
  [DELETE_TASK]: (state, {task, category}) => {
    state.tasks[category] = state.tasks[category].filter((aTask) => aTask.id !== task.id);
  },
  [ROLLBACK_STATE]: (state) => {
    Object.assign(state, JSON.parse(JSON.stringify(state.prevState)));
  },
  [BACKUP_STATE]: (state, currentState) => {
    state.prevState = JSON.parse(JSON.stringify(currentState));
  }
}
