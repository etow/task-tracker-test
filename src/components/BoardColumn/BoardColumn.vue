<template>
  <div class="board__column">
    <div class="board__column__header" :style="{ backgroundColor: getCategoryColor(category) }">
      {{ category }}
    </div>
    <div class="board__column__body">
      <Draggable
        v-model="tasks"
        group="tasks"
        item-key="id">
        <template #item="{ element }">
          <Task
            :key="element.id"
            :task="element"
            @click="setTaskToEdit(element)" />
        </template>
        <template #footer>
          <div class="board__column__footer">
            <CreateNewTask :category="category" />
          </div>
        </template>
      </Draggable>
    </div>
  </div>
</template>

<script>
import Draggable from 'vuedraggable';
import { ElMessage } from 'element-plus'
import Task from '../Task/Task.vue';
import CreateNewTask from '../CreateNewTask/CreateNewTask.vue';
import { SET_TASK_TO_EDIT, UPDATE_TASKS, ROLLBACK_STATE } from '../../store/constants';


export default {
  props: {
    category: {
      type: String,
      required: true,
    },
  },
  components: {
    Task,
    CreateNewTask,
    Draggable,
  },
  computed: {
    categories() {
      return this.$store.state.categories;
    },
    tasks: {
      get() {
        return this.$store.state.tasks[this.category];
      },
      set(tasks) {
        this.$store.dispatch(UPDATE_TASKS, { tasks, category: this.category })
          .catch((err) => {
            console.error(err.message);
            ElMessage({
              message: 'Something went wrong while trying to update the task, please try again',
              type: 'error',
              duration: 5000,
            });
            this.$store.dispatch(ROLLBACK_STATE);
          });
      },
    },
  },
  methods: {
    getCategoryColor(category) {
      return this.categories[category].color;
    },
    setTaskToEdit(task) {
      this.$store.dispatch(SET_TASK_TO_EDIT, task)
        .catch((err) => {
          console.error(err.message);
          ElMessage({
            message: 'Something went wrong while trying to edit the task, please try again',
            type: 'error',
            duration: 5000,
          });
          this.$store.dispatch(ROLLBACK_STATE);
        });
    },
  },
};
</script>

<style lang="scss" scoped>
.board__column {
  position: relative;
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  min-width: 200px;

  &:nth-child(2n+1) {
    background-color: rgba(0, 0, 0, 0.02);
  }
  &__header {
    padding: 20px;
    color: white;
    background-color: #000;
  }

  &__body {
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    overflow-y: scroll;
    padding: 20px;
    > div { // draggable area
      display: flex;
      flex-direction: column;
      flex-grow: 1;
    }
  }

  .sortable-ghost {
    cursor: grabbing;
  }

  .task {
    margin-bottom: 20px;
  }
}
</style>
