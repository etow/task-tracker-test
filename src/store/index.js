import { createStore } from 'vuex';

import mutations from './mutations';
import actions from './actions';

export default createStore({
  state: {
    tasks: {},
    categories: {},
    taskToEdit: {},
    prevState: {},
  },
  mutations,
  actions,
});
