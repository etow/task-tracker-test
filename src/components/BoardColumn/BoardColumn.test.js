import { shallowMount, flushPromises } from '@vue/test-utils';
import { createStore } from 'vuex';
import BoardColumn from './BoardColumn.vue';
import { ElMessage } from 'element-plus'
import { SET_TASK_TO_EDIT, ROLLBACK_STATE, UPDATE_TASKS } from '../../store/constants';

jest.mock('element-plus', () => ({
  ElMessage: jest.fn(),
}));

const tasks = [
  { name: 'test1', category: 'Planned', order: 0, },
  { name: 'test2', category: 'Planned', order: 1, },
  { name: 'test3', category: 'Planned', order: 2 },
];

describe('BoardColumn.vue', () => {
  it('renders correctly', async () => {
    // mock store
    const [store] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    await flushPromises();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('calculates new order of item after drag', async () => {
    const [store, { actions }] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    // change order
    const tasksUpdaded = [];
    tasksUpdaded[0] = tasks[2];
    tasksUpdaded[2] = tasks[0];

    wrapper.vm.tasks = tasksUpdaded;
    expect(actions[UPDATE_TASKS]).toHaveBeenCalledWith(expect.anything(), { tasks: tasksUpdaded, category: 'Planned' });
  });

  it('triggers set task to edit', async () => {
    const [store, { actions }] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    wrapper.vm.setTaskToEdit(tasks[0]);
    expect(actions[SET_TASK_TO_EDIT]).toHaveBeenCalledWith(expect.anything(), tasks[0]);
  });

  it('handless error when setting a task to edit', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    const [store, { actions }] = mockStore({ set_task_to_edit: jest.fn(() => Promise.reject('error'))});
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    wrapper.vm.setTaskToEdit(tasks[0]);

    expect(actions[SET_TASK_TO_EDIT]).toHaveBeenCalledWith(expect.anything(), tasks[0]);
    await flushPromises();

    expect(console.error).toHaveBeenCalled();
    expect(ElMessage).toHaveBeenCalledWith({
      message: 'Something went wrong while trying to edit the task, please try again',
      type: 'error',
      duration: 5000,
    });
    expect(actions[ROLLBACK_STATE]).toHaveBeenCalled();
    console.error.mockRestore();
  });

  it('it handles error when updating a task', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});

    const [store, { actions }] = mockStore({ update_tasks: jest.fn(() => Promise.reject('error'))});
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    // change order
    const tasksUpdaded = [];
    tasksUpdaded[0] = tasks[2];
    tasksUpdaded[2] = tasks[0];

    wrapper.vm.tasks = tasksUpdaded;

    expect(actions[UPDATE_TASKS]).toHaveBeenCalled();
    await flushPromises();

    expect(console.error).toHaveBeenCalled();
    expect(ElMessage).toHaveBeenCalledWith({
      message: 'Something went wrong while trying to update the task, please try again',
      type: 'error',
      duration: 5000,
    });
    expect(actions[ROLLBACK_STATE]).toHaveBeenCalled();
    console.error.mockRestore();
  });
});

// create wrapper helper
const createWrapper = (decorator = (n) => n) => {
  const dataToMount = {
    propsData: {
      category: 'Planned',
    },
    global: {},
  };

  decorator(dataToMount);
  return shallowMount(BoardColumn, dataToMount);
};

// mock store helper
function mockStore({ set_task_to_edit, update_tasks, rollback_state } = {}) {
  const state = {
    categories: {
      Planned: { color: 'red' },
    },
    tasks,
  };

  const actions = {
    [SET_TASK_TO_EDIT]: set_task_to_edit || jest.fn(() => Promise.resolve()),
    [UPDATE_TASKS]: update_tasks || jest.fn(() => Promise.resolve()),
    [ROLLBACK_STATE]: rollback_state || jest.fn(() => Promise.resolve()),
  }

  const store = createStore({
    state,
    actions,
  });

  return [store, { actions }];
}
