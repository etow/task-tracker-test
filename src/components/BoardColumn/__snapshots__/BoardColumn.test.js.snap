// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardColumn.vue renders correctly 1`] = `
<div class="board__column">
  <div class="board__column__header" style="background-color: red;">Planned</div>
  <div class="board__column__body">
    <draggable-stub itemkey="id" clone="function(t){return t}" tag="div" group="tasks"></draggable-stub>
  </div>
</div>
`;
